hasmer/hasmer-cli/hasmer-cli.csproj

```diff
@@ -1,11 +1,11 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFramework>net8.0</TargetFramework>
     <RootNamespace>Hasmer.CLI</RootNamespace>
     <AssemblyName>hasmer</AssemblyName>
-    <PublishAot>true</PublishAot>
+    <PublishAot>false</PublishAot>
```

Comment:
Unfortunately the stable Docker images don't support compiling AOT. It was added 1 year ago but it's (still) only [available in the nightly version](https://github.com/dotnet/dotnet-docker/issues/4129#issuecomment-1730507787).

I see four options:

1) We disable `<PublishAot>` until the official image `mcr.microsoft.com/dotnet/sdk:8.0` supports AOT. This is the approach I've taken for now.

2) We switch the base image to `mcr.microsoft.com/dotnet/nightly`, with the risk that it may break in the future. We could pin the hash of the image to mitigate this but I'm uncomfortable at the idea of using a random nightly that may or may not contain bugs.

3) We duplicate this file so that there is a flavor with `<PublishAot>true</PublishAot>` and another with `<PublishAot>false</PublishAot>`. Looks a bit bothersome to me and I'm not familiar enough with .NET to know how

4) We add a line in the Dockerfile to monkey patch on-the-fly that file, replacing `<PublishAot>true</PublishAot>` with `<PublishAot>false</PublishAot>`. I like this solution even less, I'm not fond of dynamic building shenanigans unless it's really needed.

-------

How big of a size increase/performance hit does disabling AOT incur? If it's negligible (or fast enough that we don't care) then I'm not sure it's worth handling exceptions here. In this case I'd be in favor of option 1.

What do you think?

Reply:
@devnoname120, check out the .NET AOT Dockerfile sample. It covers how you can build and deploy an AOT image without any special AOT images. It comes down to installing `clang` and `zlib1g-dev` packages in the SDK image before you start the build. Let me know if that works in this scenario.

https://github.com/dotnet/dotnet-docker/blob/08e0e7c3345a75af15ca0c1733d4f07676e0997d/samples/releasesapi/Dockerfile.debian

Reply:
@lbussell Thanks, I confirm that the AOT compiles successfully just by adding these two lines at the top:
```Dockerfile
RUN apt-get update
RUN apt-get install -y clang zlib1g-dev
```

Now the program crashes on start unfortunately:

```
docker run --volume /Users/paul/dev/unpacked/assets/:/assets hasmer disassemble -i index.android.bundle

Unhandled Exception: System.InvalidOperationException: Type Hasmer.CLI.Program+DisassembleOptions appears to be immutable, but no constructor found to accept values.
   at CommandLine.Infrastructure.ReflectionHelper.CreateDefaultImmutableInstance(Type, Type[]) + 0x134
   at CommandLine.Core.InstanceBuilder.<>c__0`1.<Build>b__0_0(Func`1 f) + 0x18
   at CommandLine.Core.InstanceBuilder.Build[T](Maybe`1, Func`3, IEnumerable`1, StringComparer, Boolean, CultureInfo, Boolean, Boolean, IEnumerable`1) + 0xfc
   at CommandLine.Core.InstanceChooser.<>c__DisplayClass0_0.<Choose>b__1() + 0xdc
   at CommandLine.Parser.ParseArguments(IEnumerable`1, Type[]) + 0xa8
   at Hasmer.CLI.Program.Main(String[] args) + 0x60
   at hasmer!<BaseAddress>+0x93b208
```

Note that if I disable AOT by changing `<PublishAot>true</PublishAot>` to `<PublishAot>false</PublishAot>` in the file `hasmer/hasmer-cli/hasmer-cli.csproj` then the crash on start doesn't happen anymore.

@lucasbaizer2 Can you confirm that hasmer actually works with AOT enabled in your local install? IMO it would be easier to just disable it, unless there are massive performance benefits but I doubt that for our use case it will be much more than just reduced image size.

Reply:
Yeah hasmer definitely isn't compatible with AOT at all so it should be disabled.
See: https://github.com/lucasbaizer2/hasmer/pull/6#issuecomment-2397527365